meadow-client/src/Blockly.purs

```diff
@@ -0,0 +1,203 @@
+  writeImpl (Dropdown fields) = JSON.write $ Record.insert type_ "field_dropdown" fields
+  writeImpl (Checkbox fields) = JSON.write $ Record.insert type_ "field_checkbox" fields
+  writeImpl (Colour fields) = JSON.write $ Record.insert type_ "field_colour" fields
+  writeImpl (Number fields) = JSON.write $ Record.insert type_ "field_number" fields
+  writeImpl (Angle fields) = JSON.write $ Record.insert type_ "field_angle" fields
+  writeImpl (Variable fields) = JSON.write $ Record.insert type_ "field_variable" fields
+  writeImpl (Date fields) = JSON.write $ Record.insert type_ "field_date" fields
+  writeImpl (Label fields) = JSON.write $ Record.insert type_ "field_label" fields
+  writeImpl (Image fields) = JSON.write $ Record.insert type_ "field_image" fields
+  writeImpl (Value fields) = JSON.write $ Record.insert type_ "input_value" fields
+  writeImpl (Statement fields) = JSON.write $ Record.insert type_ "input_statement" fields
+  writeImpl DummyRight = JSON.write $ {type: "input_dummy", align: Right}
+  writeImpl DummyLeft = JSON.write $ {type: "input_dummy", align: Left}
+  writeImpl DummyCentre = JSON.write $ {type: "input_dummy", align: Centre}
+
+data AlignDirection
+  = Left
+  | Centre
+  | Right
+
+instance writeForeignAlignDirection :: WriteForeign AlignDirection where
+  writeImpl Left = JSON.write "LEFT"
+  writeImpl Centre = JSON.write "CENTRE"
+  writeImpl Right = JSON.write "RIGHT"
+
+type BasicBlockDefinition r
+  = ( message0 :: String
+  , args0 :: Array Arg
+  , lastDummyAlign0 :: AlignDirection
+  , colour :: String
+  , fieldValue :: Maybe Pair
+  , helpUrl :: String
+  , inputsInline :: Maybe Boolean
+  , nextStatement :: Maybe String
+  , output :: Maybe String
+  , previousStatement :: Maybe String
+  , tooltip :: Maybe String
+  , extensions :: Array String
+  , mutator :: Maybe String
+  | r
+  )
+
+newtype BlockDefinition
+  = BlockDefinition (Record (BasicBlockDefinition (type :: String)))
+
+derive instance newtypeBlockDefinition :: Newtype BlockDefinition _
+
+instance writeForeignBlockDefinition :: WriteForeign BlockDefinition where
+  writeImpl (BlockDefinition fields) = JSON.write fields
+
+defaultBlockDefinition ::
+  { extensions :: Array String
+  , lastDummyAlign0 :: AlignDirection
+  , args0 :: Array Arg
+  , fieldValue :: Maybe Pair
+  , helpUrl :: String
+  , inputsInline :: Maybe Boolean
+  , mutator :: Maybe String
+  , nextStatement :: Maybe String
+  , output :: Maybe String
+  , previousStatement :: Maybe String
+  , tooltip :: Maybe String
+  }
+defaultBlockDefinition =
+  { fieldValue: Nothing
+  , lastDummyAlign0: Left
+  , args0: []
+  , helpUrl: ""
+  , inputsInline: Just true
+  , nextStatement: Nothing
+  , output: Nothing
+  , previousStatement: Nothing
+  , tooltip: Nothing
+  , extensions: []
+  , mutator: Nothing
+  }
+
+xml :: forall p i. Node (id :: String, style :: String) p i
+xml props children = element (ElemName "xml") props children
```

Comment:
_Personally_ I think these read better if you eta reduce them.

Reply:
me too!